### PR TITLE
feat: `MetaFunction` type infers `data` and `parentsData` types from loaders

### PR DESCRIPTION
**Pull request by pcattori**
Supercedes #928

See usage example in new jsdoc for `MetaFunction` or in the changeset

- [x] Docs
- [ ] Tests

Testing Strategy: I tried out the example shown in the jsdoc locally

### CODE REVIEW

File: packages/remix-server-runtime/routeModules.ts [hunk 1]
```diff
@@ -78,11 +79,65 @@ export interface LoaderFunction {
...
+  Loader extends LoaderFunction | unknown = unknown,
+  ParentsLoaders extends Record<string, LoaderFunction> = {}
+> {
   (args: {
-    data: AppData;
-    parentsData: RouteData;
+    data: Loader extends LoaderFunction ? SerializeFrom<Loader> : AppData;
+    parentsData: {
+      [k in keyof ParentsLoaders]: SerializeFrom<ParentsLoaders[k]>;
+    } & RouteData;
```
Comment (pcattori) [inline packages/remix-server-runtime/routeModules.ts:140]: `AppData` and `RouteData` including for backwards compatibility. In v2, we should change `AppData` to `unknown` and remove the need for `AppData` and `RouteData` here:

```ts
// In Remix v2:
export interface MetaFunction<
  Loader extends LoaderFunction,
  ParentsLoaders extends Record<string, LoaderFunction> = {}
> {
  (args: {
    data: SerializeFrom<Loader>;
    parentsData: {
      [k in keyof ParentsLoaders]: SerializeFrom<ParentsLoaders[k]>;
    };
   ...
```

**Review by chaance — APPROVED** [review]
LGTM! Since this is technically new API, wanna get one more set of eyes before merging?

**Review by kentcdodds — APPROVED** [review]
Holy crap I love this 👍

**Comment by kentcdodds** [comment]
@pcattori could you add some docs on this in the MetaFunction docs?

**Pull request merged**